File: reference/oracle/rng.mjs
// xoshiro128** PRNG oracle — implementation of docs/rng.md.
//
// Seeded from four literal u32 words (never expanded from a single seed); see
// docs/rng.md for why the state stays bit-exact in float64 without Math.imul
// or BigInt. Every operation that can exceed 32 bits is masked with `>>> 0`.

const u32 = (x) => x >>> 0;

// 32-bit rotate left. The `>>> 0` on the left operand keeps the shifted value
// 32 bits wide before the or; the final `>>> 0` normalizes the result.
const rotl = (x, k) => (((x >>> 0) << k) | (x >>> (32 - k))) >>> 0;

/**
 * Create a fresh stream from four u32 state words (s0..s3).
 * State must never be all-zero — a zero seed is a caller bug, not a case
 * this spec handles gracefully.
 */
export function createRng(seed) {
  if (!Array.isArray(seed) || seed.length !== 4) {
    throw new TypeError('seed must be four u32 values [s0, s1, s2, s3]');
  }
  const s = seed.map(u32);
  if (s[0] === 0 && s[1] === 0 && s[2] === 0 && s[3] === 0) {
    throw new RangeError('rng state must not be all-zero');
  }

  // The core step from docs/rng.md. Each word is masked back to 32 bits after
  // every operation that can exceed one (`* 5`, `* 9`, `<< 9`): JS's bitwise
  // operators coerce through *signed* 32 bits, so an unmasked product above
  // 2^31 would reach xor as a negative value and corrupt the state.
  function next() {
    const result = (rotl((s[1] * 5) >>> 0, 7) * 9) >>> 0;
    const t = (s[1] << 9) >>> 0;

    s[2] = (s[2] ^ s[0]) >>> 0;
    s[3] = (s[3] ^ s[1]) >>> 0;
    s[1] = (s[1] ^ s[2]) >>> 0;
    s[0] = (s[0] ^ s[3]) >>> 0;

    s[2] = (s[2] ^ t) >>> 0;
    s[3] = rotl(s[3], 11);

    return result;
  }

  return {
    next,
    // Copy of the current state, e.g. to fill the canon `rng_state` field.
    state: () => s.slice(),
  };
}

/**
 * Uniform-ish index in [0, n) from one raw u32 draw: the top 32 bits of the
 * 64-bit product r * n. Valid in JS for n < 2^21 (2,097,152) — beyond that
 * `r * n` exceeds Number.MAX_SAFE_INTEGER and the low bits are no longer
 * exact; see docs/rng.md ("Bounded draw").
 */
export function boundedDraw(next, n) {
  if (!Number.isInteger(n) || n <= 0 || n >= 2097152) {
    throw new RangeError('boundedDraw requires an integer n in (0, 2^21)');
  }
  const r = next();
  return Math.floor((r * n) / 4294967296);
}

File: core/rng.zig
//! xoshiro128** PRNG — implementation of docs/rng.md.
//!
//! Seeded from four literal u32 words (never expanded from a single seed;
//! see docs/rng.md for why). Native u32 wrapping arithmetic replaces the JS
//! oracle's `>>> 0` masks: every operation that can exceed 32 bits (`* 5`,
//! `* 9`, `<< 9`) wraps, so rotate and xor operate on the correct 32-bit
//! value exactly as docs/rng.md's JS-exactness argument requires.
//!
//! No allocator, no libc: the state is four `u32`s owned by the caller.

const std = @import("std");

pub const Rng = struct {
    s0: u32,
    s1: u32,
    s2: u32,
    s3: u32,

    /// Seeds the stream from four u32 words (s0..s3). State must never be
    /// all-zero — a zero seed is a caller bug, not a case this spec handles
    /// gracefully (xoshiro's own precondition).
    pub fn init(seed: [4]u32) Rng {
        std.debug.assert(seed[0] != 0 or seed[1] != 0 or seed[2] != 0 or seed[3] != 0);
        return .{ .s0 = seed[0], .s1 = seed[1], .s2 = seed[2], .s3 = seed[3] };
    }

    fn rotl(x: u32, comptime k: u5) u32 {
        // std.math.rotl wraps the shift count (1 +% ~k), so it stays in-range
        // even when 32 - k would overflow a u5 shift operand.
        return std.math.rotl(u32, x, k);
    }

    /// The core step from docs/rng.md: the reference xoshiro128** step
    /// verbatim, no modification.
    pub fn next(self: *Rng) u32 {
        const result = rotl(self.s1 *% 5, 7) *% 9;
        const t = self.s1 << 9;

        self.s2 ^= self.s0;
        self.s3 ^= self.s1;
        self.s1 ^= self.s2;
        self.s0 ^= self.s3;

        self.s2 ^= t;
        self.s3 = rotl(self.s3, 11);

        return result;
    }

    /// Uniform-ish index in [0, n): the top 32 bits of the 64-bit product
    /// `r * n` (docs/rng.md "Bounded draw"). Zig has real 64-bit integers,
    /// so unlike the JS oracle this has no exactness domain below 2^64;
    /// `n > 0` is all that is required. The modulo bias is the accepted,
    /// deliberate divergence documented in docs/rng.md.
    pub fn boundedDraw(self: *Rng, n: u32) u32 {
        std.debug.assert(n > 0);
        const r: u64 = self.next();
        return @truncate((r * @as(u64, n)) >> 32);
    }

    /// Current state words, e.g. to fill the canon `rng_state` field
    /// (docs/canonical-state.md).
    pub fn state(self: Rng) [4]u32 {
        return .{ self.s0, self.s1, self.s2, self.s3 };
    }
};

// docs/rng.md "Test vectors": seed [1,2,3,4], first 8 raw next() outputs —
// the same sequence reference/oracle/rng.mjs produces (AC#2).
test "next matches docs/rng.md published vector" {
    var rng = Rng.init(.{ 1, 2, 3, 4 });
    const expected = [_]u32{
        0x00002d00, 0x00000000, 0x005a7080, 0x04389d80,
        0x79199d9b, 0x61963b24, 0x4cb9b57a, 0xde9d7431,
    };
    for (expected) |want| {
        try std.testing.expectEqual(want, rng.next());
    }
    try std.testing.expectEqualSlices(
        u32,
        &[_]u32{ 0x3320a290, 0xebdc5e1d, 0x90c43618, 0xe4b42f08 },
        &rng.state(),
    );
}

// Same seed, boundedDraw(next, 573) for the first 5 draws, fresh stream
// (docs/rng.md "Test vectors").
test "boundedDraw matches docs/rng.md published vector" {
    var rng = Rng.init(.{ 1, 2, 3, 4 });
    const expected = [_]u32{ 0, 0, 0, 9, 271 };
    for (expected) |want| {
        try std.testing.expectEqual(want, rng.boundedDraw(573));
    }
}
